akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSpec.scala

```diff
@@ -569,39 +569,67 @@ class InterpreterSpec extends AkkaSpec with GraphInterpreterSpecKit {
 
   }
 
-  private[akka] case class Doubler[T]() extends PushPullStage[T, T] {
-    var oneMore: Boolean = false
-    var lastElem: T = _
-
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      oneMore = true
-      ctx.push(elem)
-    }
+  private[akka] final case class Doubler[T]() extends GraphStage[FlowShape[T, T]] {
+    val out: Outlet[T] = Outlet("Doubler.out")
+    val in: Inlet[T] = Inlet("Doubler.in")
+
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        var latest: T = _
+        var oneMore = false
+
+        override def onPush(): Unit = {
+          latest = grab(in)
+          oneMore = true
+          push(out, latest)
+        }
+
+        /**
+         * Called when the output port has received a pull, and therefore ready to emit an element, i.e. [[GraphStageLogic.push()]]
+         * is now allowed to be called on this port.
+         */
+        override def onPull(): Unit = {
+          if (oneMore) {
+            push(out, latest)
+            oneMore = false
+          } else {
+            pull(in)
+          }
+        }
+
+        setHandlers(in, out, this)
+      }
 
-    override def onPull(ctx: Context[T]): SyncDirective = {
-      if (oneMore) {
-        oneMore = false
-        ctx.push(lastElem)
-      } else ctx.pull()
-    }
+    override val shape: FlowShape[T, T] =  FlowShape(in, out)
   }
 
-  private[akka] case class KeepGoing[T]() extends PushPullStage[T, T] {
-    var lastElem: T = _
+  private[akka] final case class KeepGoing[T]() extends GraphStage[FlowShape[T, T]] {
+    val in = Inlet[T]("KeepGoing.in")
+    val out = Outlet[T]("KeepGoing.out")
 
-    override def onPush(elem: T, ctx: Context[T]): SyncDirective = {
-      lastElem = elem
-      ctx.push(elem)
-    }
+    override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+      new GraphStageLogic(shape) with InHandler with OutHandler {
+        var lastElem: T = _
 
-    override def onPull(ctx: Context[T]): SyncDirective = {
-      if (ctx.isFinishing) {
-        ctx.push(lastElem)
-      } else ctx.pull()
-    }
+        override def onPush(): Unit = {
+          lastElem = grab(in)
+          push(out, lastElem)
+        }
+
+        override def onPull(): Unit = {
+          if (isClosed(in)) {
+            push(out, lastElem)
```

Comment:
this does not handle the case where `lastElem` is uninitialized. It seems like though neither did the original and this seems to be like test code so maybe adding a comment somewhere is enough.